src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,2059 @@
+
+            eps_i = self.config.eps / lprobs.size(-1)
+            loss = (1.0 - self.config.eps) * loss + eps_i * smooth_loss
+
+        return loss
+
+    def prepare_inputs_for_generation(
+        self, decoder_input_ids, past, attention_mask, use_cache, encoder_outputs, **kwargs
+    ):
+        assert encoder_outputs is not None, "`encoder_outputs` have to be passed for generation."
+
+        if past:
+            decoder_input_ids = decoder_input_ids[:, -1:]
+        # first step, decoder_cached_states are empty
+        return {
+            "input_ids": None,  # encoder_outputs is defined. input_ids not needed
+            "encoder_outputs": encoder_outputs,
+            "past_key_values": past,
+            "decoder_input_ids": decoder_input_ids,
+            "attention_mask": attention_mask,
+            "use_cache": use_cache,
+        }
+
+    def prepare_logits_for_generation(self, logits, cur_len, max_length):
+        if (cur_len == 1) or (cur_len == max_length - 1 and self.config.eos_token_id is not None):
+            return self._force_token_ids_generation(logits, self.config.bos_token_id)
+        return logits
+
+    def _force_token_ids_generation(self, scores, token_ids) -> None:
+        """force one of token_ids to be generated by setting prob of all other tokens to 0"""
+        if isinstance(token_ids, int):
+            token_ids = [token_ids]
+        all_but_token_ids_mask = torch.tensor(
+            [x for x in range(self.config.vocab_size) if x not in token_ids],
+            dtype=torch.long,
+            device=next(self.parameters()).device,
+        )
+        assert len(scores.shape) == 2, "Scores should be of rank 2"
+        scores[:, all_but_token_ids_mask] = -float("inf")
+        return scores
+
+    @staticmethod
+    def _reorder_cache(past, beam_idx):
+        # this function reorders the cache for beam search
+        def _reorder_cache(cache_dict, beam_idx):
+            for k, key_value_states in cache_dict.items():
+                if key_value_states is not None:
+                    cache_dict[k] = key_value_states.index_select(0, beam_idx)
+            return cache_dict
+
+        reordered_past = []
+        for layer_past in past:
+            # get the correct batch idx from decoder layer's batch dim for cross and self-attn
+            layer_past_new = {
+                attn_key: _reorder_cache(attn_cache, beam_idx) for attn_key, attn_cache in layer_past.items()
+            }
+            reordered_past.append(layer_past_new)
+        return reordered_past
+
+    def get_encoder(self):
+        return self.prophetnet.encoder
+
+    def get_decoder(self):
+        return self.prophetnet.decoder
+
+
+@add_start_docstrings(
+    "The standalone decoder part of the ProphetNetModel with a lm head on top. The model can be used for causal language modeling.",
+    PROPHETNET_START_DOCSTRING,
+)
+class ProphetNetForCausalLM(ProphetNetPreTrainedModel):
```

Comment:
I don't understand why we need this class.
Can't we just compute loss in `ForConditionalGeneration` if labels are specified.

Reply:
This class makes it possible to use ProphetNet as a "standalone" decoder model, like GPT2